R/counts.R

```diff
@@ -196,62 +188,91 @@ get_single_cell_experiment <- function(
+    
     if (length(cells) < nrow(df)){
-        single_line_str(
-            "The number of cells in the SingleCellExperiment will be less than the
-            number of cells you have selected from the metadata.
-            Are cell IDs duplicated? Or, do cell IDs correspond to the counts file?
-            "
-        ) |> cli_alert_warning()
-        df <- filter(df, .data$cell_ %in% cells)
+      single_line_str(
+        "The number of cells in the SingleCellExperiment will be less than the
+                number of cells you have selected from the metadata.
+                Are cell IDs duplicated? Or, do cell IDs correspond to the counts file?
+                "
+      ) |> cli_alert_warning()
+      df <- filter(df, .data$cell_ %in% cells)
     }
     else if (length(cells) > nrow(df)){
-        cli_abort("This should never happen")
+      cli_abort("This should never happen")
     }
     
-    # Fix for https://github.com/tidyverse/dplyr/issues/6746
-    force(i)
-    
     new_coldata <- df |>
-        # We need to make the cell names globally unique, which we can guarantee
-        # by adding a suffix that is derived from file_id_db, which is the
-        # grouping variable
-        mutate(original_cell_id = .data$cell_, cell_ = glue("{cell_}_{i}")) |>
-        column_to_rownames("cell_") |>
-        as("DataFrame")
+      mutate(original_cell_id = .data$cell_, cell_ = glue("{cell_}_{i}")) |>
+      column_to_rownames("cell_") |>
+      as("DataFrame")
     
-    `if`(
-            is.null(features),
-            sce[, new_coldata$original_cell_id],
-            {
-                # Optionally subset the genes
-                genes <- rownames(sce) |> intersect(features)
-                sce[genes, new_coldata$original_cell_id]
-            }
+    experiment <- `if`(
+      is.null(features),
+      experiment[, new_coldata$original_cell_id],
+      {
+        # Optionally subset the genes
+        genes <- rownames(experiment) |> intersect(features)
+        experiment[genes, new_coldata$original_cell_id]
+      }
     ) |>
-        `colnames<-`(new_coldata$cell_) |>
-        `colData<-`(value = new_coldata)
+      `colnames<-`(new_coldata$cell_) |>
+      `colData<-`(value = new_coldata)
+  }
+  else if (type == "pseudobulk") {
+    # remove cell-level annotations
+    cell_level_anno <- c("cell_", "cell_type", "confidence_class", "file_id_db",
+                         "cell_annotation_blueprint_singler",
+                         "cell_annotation_monaco_singler", 
+                         "cell_annotation_azimuth_l2",
+                         "cell_type_ontology_term_id",
+                         "sample_id_db")
+    
+    new_coldata <- df |> select(-dplyr::all_of(cell_level_anno)) |> distinct() |>
+      mutate(sample_identifier = glue("{sample_}___{cell_type_harmonised}"),
+             original_sample_id = .data$sample_identifier) |>
+      column_to_rownames("original_sample_id")
```

Comment:
This is just for readability.
```suggestion
    new_coldata <- df |>
        select(-dplyr::all_of(cell_level_anno)) |>
        distinct() |>
        mutate(
            sample_identifier = glue("{sample_}___{cell_type_harmonised}"),
            original_sample_id = .data$sample_identifier
        ) |>
        column_to_rownames("original_sample_id")
```